helper/max_length_slice.go

```diff
@@ -0,0 +1,51 @@
+package helper
+
+import "sync"
+
+type MaxLengthStringSliceAdaptor struct {
+	s      []string
+	maxlen int
+	lock   sync.Mutex
+}
+
+// Create an adaptor for given string slice
+func NewMaxLengthStringSliceAdaptor(s []string, maxlen int) *MaxLengthStringSliceAdaptor {
+	return &MaxLengthStringSliceAdaptor{
+		s:      s,
+		maxlen: maxlen,
+	}
+}
+
+func (m MaxLengthStringSliceAdaptor) MaxLen() int {
+	return m.maxlen
+}
+
+func (m *MaxLengthStringSliceAdaptor) Put(str string) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.s = append(m.s, str)
+	if len(m.s) > m.MaxLen() {
+		m.s = m.s[1:]
```

Comment:
Don't do this, make the slice cycle instead of create new slice. The golang's slice reuses an inner array of the original, that is it works much like the `substring` method of Java before 1.8. The memory may not be released correctly as your expect!